src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -4538,7 +4544,48 @@ public override IOperation VisitLiteral(ILiteralOperation operation, int? captur
         {
             return new LiteralExpression(semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
         }
-        
+
+        public override IOperation VisitLocalReference(ILocalReferenceOperation operation, int? captureIdForResult)
+        {
+            return new LocalReferenceExpression(operation.Local, operation.IsDeclaration, semanticModel: null, operation.Syntax,
+                                                operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        public override IOperation VisitParameterReference(IParameterReferenceOperation operation, int? captureIdForResult)
+        {
+            return new ParameterReferenceExpression(operation.Parameter, semanticModel: null, operation.Syntax,
+                                                    operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        public override IOperation VisitFieldReference(IFieldReferenceOperation operation, int? captureIdForResult)
+        {
+            IOperation visitedInstance = operation.Field.IsStatic ? null : Visit(operation.Instance);
+            return new FieldReferenceExpression(operation.Field, operation.IsDeclaration, visitedInstance, semanticModel: null,
+                                                operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        public override IOperation VisitMethodReference(IMethodReferenceOperation operation, int? captureIdForResult)
+        {
+            IOperation visitedInstance = operation.Method.IsStatic ? null : Visit(operation.Instance);
+            return new MethodReferenceExpression(operation.Method, operation.IsVirtual, visitedInstance, semanticModel: null,
+                                                 operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        public override IOperation VisitPropertyReference(IPropertyReferenceOperation operation, int? captureIdForResult)
+        {
+            IOperation instance = operation.Property.IsStatic ? null : operation.Instance;
```

Comment:
Consider just modifying VisitInstanceWithArguments to take an ISymbol, and doing the check there, instead of duplicating here and in VisitInvocation. You could also consider having it accept an empty or default array of arguments, and reusing it for visiting all instances, so we can have the static check in one place for all.

Reply:
I considered that initially, but given the fact that the static check is pretty trivial for callers, `VisitInstanceWithArguments` is currently used only at 2 places, and it has a much cleaner API to implementation mapping, I chose the current structure. Let me know if you feel strongly about your suggestion and I can refactor in my next commit (if there is additional feedback) or next PR.

---
In reply to: [188427893](https://github.com/dotnet/roslyn/pull/26797#discussion_r188427893) [](ancestors = 188427893)

Reply:
Well, I also meant to consider using VisitInstanceWithArguments in more places as well, ie make it work for VisitEventReference, VisitMethodReference, etc. I do think it would be good to have one location where these checks are done so it's easier to verify, a follow up is fine.

---
In reply to: [188429842](https://github.com/dotnet/roslyn/pull/26797#discussion_r188429842) [](ancestors = 188429842,188427893)

Reply:
> I also meant to consider using VisitInstanceWithArguments in more places as well, ie make it work for VisitEventReference, VisitMethodReference, etc.

That sounds a bit convoluted to me, especially since they can never have arguments. I think your approach would be more valuable if the shared code grows more than one line: `IOperation visitedInstance = operation.Event.IsStatic ? null : Visit(operation.Instance);`

---
In reply to: [188431334](https://github.com/dotnet/roslyn/pull/26797#discussion_r188431334) [](ancestors = 188431334,188429842,188427893)